File: ios-catalogue/App/Shell/ContrastBadge.swift
import SwiftUI
import CoreUI
#if canImport(UIKit)
import UIKit
#endif

/// Live WCAG contrast indicator. Computes the actual ratio between the
/// active theme's primary text and base surface, then surfaces it as a
/// coloured chip — green at ≥ 4.5:1, amber 3-4.5:1, red below.
///
/// Recomputes when the theme flips so the user can sanity-check overrides
/// (e.g. an in-progress dark theme) at a glance.
struct ContrastBadge: View {
    @Environment(\.colorScheme) private var scheme

    var body: some View {
        let textColor = PrismaSemanticColors.textPrimary.themed(scheme)
        let surfaceColor = PrismaSemanticColors.surfaceBase.themed(scheme)
        let ratio = contrastRatio(fg: textColor, bg: surfaceColor)
        let (dotColor, label): (Color, String) = {
            switch ratio {
            case 7.0...:        return (PrismaSemanticColors.statusSuccessDefault.themed(scheme), "AAA")
            case 4.5..<7.0:     return (PrismaSemanticColors.statusSuccessDefault.themed(scheme), "AA")
            case 3.0..<4.5:     return (PrismaSemanticColors.statusWarningDefault.themed(scheme), "AA-")
            default:            return (PrismaSemanticColors.statusDangerDefault.themed(scheme), "FAIL")
            }
        }()
        let ratioStr = String(format: "%.1f:1", ratio)

        HStack(spacing: PrismaSpacing.sp2) {
            Circle().fill(dotColor).frame(width: 8, height: 8)
            Text(label)
                .font(PrismaTypography.labelSm.font)
                .foregroundStyle(PrismaSemanticColors.textSecondary.themed(scheme))
            Text(ratioStr)
                .font(PrismaTypography.labelSm.font)
                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
        }
        .padding(.horizontal, PrismaSpacing.sp3)
        .padding(.vertical, 6)
        .background(PrismaSemanticColors.surfaceRaised.themed(scheme))
        .overlay(
            Capsule().strokeBorder(PrismaSemanticColors.borderSubtle.themed(scheme), lineWidth: 1)
        )
        .clipShape(Capsule())
        .accessibilityElement(children: .combine)
        .accessibilityLabel("WCAG \(label), contrast \(ratioStr)")
    }
}

private func contrastRatio(fg: Color, bg: Color) -> Double {
    let l1 = relativeLuminance(fg)
    let l2 = relativeLuminance(bg)
    let lighter = max(l1, l2)
    let darker = min(l1, l2)
    return (lighter + 0.05) / (darker + 0.05)
}

private func relativeLuminance(_ color: Color) -> Double {
    #if canImport(UIKit)
    var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
    UIColor(color).getRed(&r, green: &g, blue: &b, alpha: &a)
    func channel(_ v: CGFloat) -> Double {
        let d = Double(v)
        return d <= 0.03928 ? d / 12.92 : pow((d + 0.055) / 1.055, 2.4)
    }
    return 0.2126 * channel(r) + 0.7152 * channel(g) + 0.0722 * channel(b)
    #else
    return 0
    #endif
}

File: android-catalogue/catalogue/src/main/java/xyz/ksharma/prisma/catalogue/shell/ContrastBadge.kt
package xyz.ksharma.prisma.catalogue.shell

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.coreui.LocalPrismaIsDark
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography
import kotlin.math.pow

/**
 * Live WCAG contrast indicator. Computes the actual ratio between the
 * active theme's primary text and base surface, then surfaces it as a
 * coloured chip — green at ≥ 4.5:1, amber 3-4.5:1, red below.
 *
 * Recomputes when the theme flips so the user can sanity-check overrides
 * (e.g. an in-progress dark theme) at a glance.
 */
@Composable
internal fun ContrastBadge(modifier: Modifier = Modifier) {
    val isDark = LocalPrismaIsDark.current
    val text = PrismaSemanticColors.TextPrimary.resolve(isDark)
    val surface = PrismaSemanticColors.SurfaceBase.resolve(isDark)
    val ratio = remember(isDark, text, surface) { contrastRatio(text, surface) }
    val (dotColor, label) = when {
        ratio >= 7.0 -> PrismaSemanticColors.StatusSuccessDefault.resolve(isDark) to "AAA"
        ratio >= 4.5 -> PrismaSemanticColors.StatusSuccessDefault.resolve(isDark) to "AA"
        ratio >= 3.0 -> PrismaSemanticColors.StatusWarningDefault.resolve(isDark) to "AA-"
        else -> PrismaSemanticColors.StatusDangerDefault.resolve(isDark) to "FAIL"
    }
    val ratioStr = "%.1f:1".format(ratio)

    Row(
        modifier = modifier
            .clip(RoundedCornerShape(PrismaRadius.Full))
            .background(PrismaSemanticColors.SurfaceRaised.resolve(isDark))
            .border(
                1.dp,
                PrismaSemanticColors.BorderSubtle.resolve(isDark),
                RoundedCornerShape(PrismaRadius.Full),
            )
            .padding(horizontal = PrismaSpacing.Sp3, vertical = 6.dp)
            .semantics {
                contentDescription = "WCAG $label, contrast $ratioStr"
            },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp2),
    ) {
        androidx.compose.foundation.layout.Box(
            modifier = Modifier
                .size(8.dp)
                .clip(CircleShape)
                .background(dotColor),
        )
        Text(
            text = label,
            style = PrismaTypography.LabelSm,
            color = PrismaSemanticColors.TextSecondary.resolve(isDark),
        )
        Text(
            text = ratioStr,
            style = PrismaTypography.LabelSm,
            color = PrismaSemanticColors.TextTertiary.resolve(isDark),
        )
    }
}

/** WCAG 2.1 relative luminance + contrast ratio. */
private fun contrastRatio(fg: Color, bg: Color): Double {
    val l1 = relativeLuminance(fg)
    val l2 = relativeLuminance(bg)
    val lighter = maxOf(l1, l2)
    val darker = minOf(l1, l2)
    return (lighter + 0.05) / (darker + 0.05)
}

private fun relativeLuminance(c: Color): Double {
    fun linear(channel: Float): Double {
        val v = channel.toDouble()
        return if (v <= 0.03928) v / 12.92 else ((v + 0.055) / 1.055).pow(2.4)
    }
    return 0.2126 * linear(c.red) + 0.7152 * linear(c.green) + 0.0722 * linear(c.blue)
}
